File: library/src/integrationTest/java/org/crispy_fish/util/ResultConditions.java
package org.crispy_fish.util;

import org.assertj.core.api.Condition;
import org.assertj.core.condition.AllOf;
import org.crispy_fish.domain.Result;
import org.crispy_fish.domain.payload.Driver;

public class ResultConditions {

    public static Condition<Result> positionEquals(final int position) {
        return new Condition<>(result -> result.position == position, "result.position == %d", position);
    }

    public static Condition<Result> driverEquals(final Driver driver) {
        return new Condition<>(result -> result.driver.equals(driver), "result.driver == %s", driver.toString());
    }

    public static Condition<Result> driverEquals(final String classing, final String number) {
        Driver driver = new Driver();
        driver.classing = classing;
        driver.number = number;
        return driverEquals(driver);
    }

    public static Condition<Result> driverFinished(final int position, final String classing, final String number) {
        return AllOf.allOf(
                positionEquals(position),
                driverEquals(classing, number)
        );
    }
}

File: library/src/test/kotlin/tech/coner/crispyfish/util/ResultConditions.kt
package tech.coner.crispyfish.util

import org.assertj.core.api.Condition
import org.assertj.core.condition.AllOf
import tech.coner.crispyfish.model.Result
import java.util.function.Predicate

object ResultConditions {

    fun positionEquals(position: Int) = Condition<Result>(
                Predicate{ result -> result.position == position },
                "result.position == %d",
                position
    )

    fun driverNumbersEqual(category: String?, handicap: String, number: String) = Condition<Result>(
            Predicate { result ->
                (result.driver.signage.classing?.category?.abbreviation ?: "").equals(category, ignoreCase = true)
                && result.driver.signage.classing?.handicap?.abbreviation.equals(handicap, ignoreCase = true)
                && result.driver.signage.number == number
            },
            "result.driver == %s",
            "${category ?: ""}$handicap $number"
    )

    fun driverNameNotNullOrEmpty(): Condition<Result> {
        return Condition(Predicate{ result ->
            !result.driver.firstName.isNullOrBlank()
            && !result.driver.lastName.isNullOrBlank()
        }, "result.driver first and last last names are not null or blank")
    }

    fun driverFinished(position: Int, category: String?, handicap: String, number: String): Condition<Result> {
        return AllOf.allOf(
                positionEquals(position),
                driverNumbersEqual(category, handicap, number)
        )
    }
}
